[bug] Coverage files getting overwritten for non --single-instrumentation-call Spoon task

psrajat opened the pull request:
The multi instrumented spoon task (where `--single-instrumentation-call` is not provided as an argument) works by running command `adb shell am instrument` [https://developer.android.com/studio/test/command-line#RunTestsDevice](link) for each test case.

If the coverage flag is enabled for spoon task, it will create a coverage file for each test case. This behavior is not causing any problems for `--single-instrumentation-call` tasks. But for multi instrumented tasks, The coverage file of previous test gets replaced when the next test runs because each file have the same name (`coverage.ec`).

This change solves the issue by pulling the file from device after each test runs and in the merging the all the *.ec files into one.

Coverage file for each test will be available to the user with a proper naming convention for debugging purposes.

Commit:
[bug]

The multi instrumented spoon task (where --single-instrumentation-call is not provided as argument) works by running command `adb shell am instrument` for each test case.

If the coverage flag is enabled for spoon task, it will create a coverage file for each test case. This behavior is not causing any problems for --single-instrumentation-call tasks. But for multi instrumented tasks, Spoon is replacing the previous test's coverage file when the next test runs because each file have the same name (`coverage.ec`).

This change solves the issue by pulling the file from device after each test runs and in the merging the all the *.ec files into one.

Coverage file for each test will be available to the user with a proper naming convention for debugging purposes.

edenman commented on spoon-runner/src/main/java/com/squareup/spoon/SpoonDeviceRunner.java:
```diff
@@ -396,6 +405,26 @@ private void pullCoverageFile(IDevice device) {
+   * @param device
+   * @param testIdentifier
+   */
+  private void pullCoverageFile(IDevice device, String testIdentifier) {
```
this feels like it should share implementation with the `pullCoverageFile` method above.  Something like:
```
private void pullCoverageFile(IDevice device) {
  doPullCoverageFile(device, COVERAGE_FILE);
}
private void pullCoverageFile(IDevice device, String testIdentifier) {
  doPullCoverageFile(device, testIdentifier + "_" + COVERAGE_FILE);
}
private void doPullCoverageFile(IDevice device, String localFileName) {
    coverageDir.mkdirs();
    File coverageFile = new File(coverageDir, localFileName;
    logInfo("Pulling Code Coverage file for %s to %s", testIdentifier,
            coverageFile.getAbsolutePath());
    String remotePath;
    try {
      remotePath = getExternalStoragePath(device, COVERAGE_FILE);
    } catch (Exception exception) {
      throw new RuntimeException("error while calculating coverage file path.", exception);
    }
    adbPullFile(device, remotePath, coverageFile.getAbsolutePath());
)

```

rajat-bb replied:
Thanks Eric for pointing this out. I have fixed this.

Also, the Travis CI job seems to be failing with an error:
`Failed to install the following Android SDK packages as some licences have not been accepted.`
Should I look into this?

edenman reviewed (changes requested):
Thanks for the PR, just one requested change

Commit:
[refactor] Remove duplicate code in pullCoverageFile() method

Commit:
Add unit test for mergeAllCoverageFiles method

Commit:
Add method descriptions

Commit:
[fix] Travis CI Build Failure fix.

The original shell script was not accepting the licenses.
Fixed this using sdkmanager to accept all the licenses.

rajat-bb commented:
Hi @edenman 
I have also fixed the Travis CI Android License Acceptance Issue.

edenman commented:
Thanks again @rajat-bb!

mdenissov commented:
Looks like the spoon-runner:2.0.0-SNAPSHOT has not been uploaded to the sonatype snapshot repository


Spoon runner:
https://oss.sonatype.org/#nexus-search;gav~com.squareup.spoon~spoon-runner~~~~kw,versionexpand
Uploaded Date:
Wed Oct 03 2018 06:43:05 GMT+0700 (Novosibirsk Standard Time)

Spoon client:
https://oss.sonatype.org/#nexus-search;gav~com.squareup.spoon~spoon-client~~~~kw,versionexpand
Uploaded Date: Fri Nov 06 2020 03:23:47 GMT+0700 (Novosibirsk Standard Time)

OlliZi commented:
Please, where can i download new the release for the spoon-runner (master)?
